src/Adapter/Order/CommandHandler/DeleteProductFromOrderHandler.php

```diff
@@ -54,22 +76,91 @@ public function handle(DeleteProductFromOrderCommand $command)
 
-        // Reinject quantity in stock
-        $this->reinjectQuantity($orderDetail, $orderDetail->product_quantity, true);
+        try {
+            $result = true;
+            $oldCartRules = $cart->getCartRules();
+            $result &= $this->updateCart($orderDetail, $cart);
+            $result &= $this->updateCartRules($order, $oldCartRules, $cart->getCartRules());
+            $result &= $this->updateOrderInvoice($orderDetail);
 
-        $result &= $this->updateOrderWeight($order);
+            // Reinject quantity in stock
+            $this->reinjectQuantity($orderDetail, $orderDetail->product_quantity, true);
 
-        if (!$result) {
-            throw new OrderException('An error occurred while attempting to delete product from order.');
+            $result &= $this->updateOrder($order, $orderDetail, $cart);
+            $result &= $this->updateOrderWeight($order);
+
+            if (!$result) {
+                throw new OrderException('An error occurred while attempting to delete product from order.');
+            }
+
+            $order = $order->refreshShippingCost();
+
+            Hook::exec('actionOrderEdited', ['order' => $order]);
+        } catch (Exception $e) {
+            $this->contextStateManager->restoreContext();
+            throw $e;
         }
 
-        $order = $order->refreshShippingCost();
+        $this->contextStateManager->restoreContext();
+    }
+
+    /**
+     * @param OrderDetail $orderDetail
+     * @param Cart $cart
+     *
+     * @return bool
+     */
+    private function updateCart(OrderDetail $orderDetail, Cart &$cart)
+    {
+        // Update Product Quantity in the cart
+        $cart->updateQty(
+            $orderDetail->product_quantity,
+            $orderDetail->product_id,
+            $orderDetail->product_attribute_id,
+            false,
+            'down'
+        );
+        // Remove & Add CartRules applied to cart
+        CartRule::autoRemoveFromCart();
+        CartRule::autoAddToCart();
 
-        Hook::exec('actionOrderEdited', ['order' => $order]);
+        return $cart->update();
+    }
+
+    /**
+     * Remove previous cart rules applied to order
+     *
+     * @param Order $order
+     * @param array $oldCartRules
+     * @param array $newCartRules
+     *
+     * @return bool
+     */
+    private function updateCartRules(Order $order, array $oldCartRules, array $newCartRules)
+    {
+        $result = array_diff(
+            array_map('serialize', $oldCartRules),
+            array_map('serialize', $newCartRules)
+        );
+        $result = array_map('unserialize', $result);
+
+        foreach ($order->getCartRules() as $orderCartRule) {
+            foreach ($result as $itemCartRule) {
+                if ($itemCartRule['id_cart_rule'] == $orderCartRule['id_cart_rule']) {
+                    $orderCartRule = new OrderCartRule($orderCartRule['id_order_cart_rule']);
+                    $orderCartRule->delete();
```

Comment:
return false if the deletion of orderCartRule returns false ?

Reply:
I dont know what's the consequence of partially deleted cart rules

Reply:
Fixed